### PR TITLE
Move out `crypto/aes`

### PR DESCRIPTION
**Pull request by florianduros**
## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md)).

Task https://github.com/element-hq/element-web/issues/26922

The goal of the new `src/utils` folder is to share utility functions without to have to import all the js-sdk content (ie: `src/matrix.ts`). See https://github.com/matrix-org/matrix-js-sdk/issues/4339

Changes:
- `decryptAES` and `encryptAES`  are moved into `src/utils`.
- `deriveKeys` is used only by `decryptAES` and `encryptAES`. Moved into `src/`to keep it internal. 
- Rename `IEncryptedPayload` to `AESEncryptedSecretStoragePayload`. Move it into `src/@types/`

The PR can be reviewed by commit.

### CODE REVIEW

File: src/utils/encryptAES.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * @param key - the encryption key to use as an input to the HKDF function which is used to derive the AES key for
+ *    encryption. Obviously, the same key must be provided when decrypting.
+ * @param name - the name of the secret. Used as an input to the HKDF operation which is used to derive the AES key,
+ *    so again the same value must be provided when decrypting.
+ * @param ivStr - the base64-encoded initialization vector to use. If not supplied, a random one will be generated.
+ *
+ * @returns The encrypted result, including the ciphertext itself, the initialization vector (as supplied in `ivStr`,
+ *   or generated), and an HMAC on the ciphertext — all base64-encoded.
+ */
+export async function encryptAES(
```
Comment (richvdh) [inline src/utils/encryptAES.ts:34]: Given the objective of the `utils` hierarchy (to expose small modules that have only a single piece of functionality), I think we should use default exports here (and for `decryptAES`), with

```
export default async function ...
```

(Obviously, you'll have to update all the imports from `import { encryptAES } from ...` to `import encryptAES from ...`.)

File: src/utils/decryptAES.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Decrypt an AES-encrypted string.
+ *
+ * @param data - the encrypted data, returned by {@link encryptAES}.
+ * @param key - the encryption key to use as an input to the HKDF function which is used to derive the AES key. Must
+ *    be the same as provided to {@link encryptAES}.
+ * @param name - the name of the secret. Also used as an input to the HKDF operation which is used to derive the AES
+ *    key, so again must be the same as provided to {@link encryptAES}.
+ */
+export async function decryptAES(data: SecretEncryptedPayload, key: Uint8Array, name: string): Promise<string> {
```
[the remark this answers is not in the document]
Comment (richvdh) [reply src/utils/decryptAES.ts:30]: Also, rename the file, of course.

File: src/utils/decryptAES.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+import { decodeBase64 } from "../base64.ts";
+import { deriveKeys } from "../rust-crypto/deriveKeys.ts";
+import { SecretEncryptedPayload } from "./@types/SecretEncryptedPayload.ts";
+
+/**
+ * Decrypt an AES-encrypted string.
+ *
+ * @param data - the encrypted data, returned by {@link encryptAES}.
```
Comment (richvdh) [inline src/utils/decryptAES.ts:24]: These links don't work. Try something like:

```suggestion
@param data - the encrypted data, returned by {@link utils/encryptAES.default | encryptAES}.
```

(Assuming you're using default exports.)

**Review by richvdh — CHANGES_REQUESTED** [review]
looks like the right direction! A few suggestions for improvements.

File: src/utils/@types/SecretEncryptedPayload.ts
[the remark this answers is not in the document]
Comment (florianduros) [reply src/utils/@types/SecretEncryptedPayload.ts:1]: we will importing all the types when we import `AESEncryptedSecretStoragePayload`.

[the remark this answers is not in the document]
Comment (richvdh) [reply src/utils/@types/SecretEncryptedPayload.ts:1]: That's probably ok though? They are only typescript types, there is no runtime cost.

**Review by richvdh — APPROVED** [review]
https://github.com/matrix-org/matrix-js-sdk/pull/4431#discussion_r1778314286 seems to have got lost.

LGTM otherwise

**Pull request merged**